File: pkg_rospy/scripts/simple_action_server_all.py
#! /usr/bin/env python

# Counts until count_upto with wait_time between each count, sends feedback after 
# each count and sends the result on completion

# reference: https://docs.ros.org/en/melodic/api/actionlib/html/classactionlib_1_1simple__action__server_1_1SimpleActionServer.html

import rospy
import actionlib
from pkg_rospy.msg import CountNumbersAction, CountNumbersGoal, \
        CountNumbersResult, CountNumbersFeedback

class CountNumbersServer:
    def __init__(self):
        self._as = actionlib.SimpleActionServer('/count_numbers', CountNumbersAction, \
                execute_cb = self.on_goal, auto_start=False)
        self._as.start()
        rospy.loginfo("count numbers server started")

        self._counter = 0
    
    def on_goal(self, goal):
        rospy.loginfo("Goal has be en received")
        rospy.loginfo(goal)
        count_upto = goal.count_upto
        wait_time = goal.wait_time

        self._counter = 0
        rate = rospy.Rate(1.0/wait_time)

        feedback = CountNumbersFeedback()
        some_error_occoured = False
        preempt_reuested = False
        while (not rospy.is_shutdown()) and (self._counter < count_upto):
            if some_error_occoured:     #counting has failed due to some error
                break
            if self._as.is_preempt_requested():
                preempt_reuested = True
                break

            self._counter += 1

            feedback.current_count = self._counter
            self._as.publish_feedback(feedback)
            
            rate.sleep()
        
        result = CountNumbersResult()
        result.final_count = self._counter
        
        if some_error_occoured:
            self._as.set_aborted(result)
            rospy.logwarn("Goal aborted")
        elif preempt_reuested:
            self._as.set_preempted(result)
            rospy.logwarn("Goal Preempted")
        else:
            self._as.set_succeeded(result)
            rospy.loginfo("Goal completed succesfully")

def main():
    rospy.init_node("count_numbers_server")
    Server = CountNumbersServer()
    rospy.spin()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: pkg_roscpp/src/simple_action_server_all.cpp
// reference: https://docs.ros.org/en/melodic/api/actionlib/html/classactionlib_1_1SimpleActionServer.html

#include "ros/ros.h"
#include "actionlib/server/simple_action_server.h"
#include "pkg_roscpp/CountNumbersAction.h"

class ActionServer{
protected:
    ros::NodeHandle nh_;
    actionlib::SimpleActionServer<pkg_roscpp::CountNumbersAction> as_;

    pkg_roscpp::CountNumbersFeedback feedback_;
    pkg_roscpp::CountNumbersResult result_;

public:
    ActionServer(std::string name) : 
            as_(nh_, name, boost::bind(&ActionServer::executeCB, this, _1), false)
    {
        as_.start();
        ROS_INFO("Action Server started");
    }

    ~ActionServer(void){}

    void executeCB(const pkg_roscpp::CountNumbersGoalConstPtr &goal){
        int32_t count = 0;
        int32_t count_upto = goal->count_upto;
        float wait_time = goal->wait_time;
        ROS_INFO("Goal received, count_upto: [%d], wait_time: [%f]", count_upto, wait_time);

        ros::Rate rate(1.0/wait_time);

        bool some_error_occoured = false;
        bool preempt_reuested = false;
        pkg_roscpp::CountNumbersFeedback feedback_;

        while (!ros::isShuttingDown() && count<count_upto){
            if (some_error_occoured)
                break;
            if (as_.isPreemptRequested()){
                preempt_reuested = true;
                break;
            }
            
            count++;
            feedback_.current_count = count;
            as_.publishFeedback(feedback_);

            rate.sleep();
        }
        result_.final_count = count;

        if (some_error_occoured){
            as_.setAborted(result_);
            ROS_WARN("Goal aborted");
        }
        else if(preempt_reuested){
            as_.setPreempted(result_);
            ROS_WARN("Goal preempted");
        }
        else{
            as_.setSucceeded(result_);
            ROS_INFO("Goal completed succesfully");
        }
    }
};

int main(int argc, char **argv){
    ros::init(argc, argv, "simple_action_server_all");
    ActionServer Server("count_numbers");

    ros::spin();

    return 0;
}